File: src/Fuaran.UI.VisualBasic.Conformance.Tests/Coverage.vb
Imports System.Linq
Imports Microsoft.FSharp.Reflection
Imports Fuaran.UI.VisualBasic
' `Fuaran.UI.Types.NodeKind` is a type ABBREVIATION since the 692 swap — erased
' in metadata, so VB reflects the generated declaration directly.
Imports FsTypes = Fuaran.UI.Generated

' Phase 311 mapping-completeness guard — the §11 forward-coupling reminder for the VB
' translator: enumerate every shipped NodeKind/spec case by reflecting the F# DUs and
' assert a VB XML element maps each. A new F# kind with no VB mapping fails here.
Module Coverage

    ' The four NodeKind cases that wrap a sub-kind DU (Layout of LayoutKind, …);
    ' their leaf kinds are enumerated from those sub-DUs. Every OTHER top-level
    ' NodeKind case (Custom / ErrorBoundary / FragmentDecl / FragmentRef / Mount /
    ' any future one) authors under its own case name — derived by reflection so a
    ' new top-level NodeKind case can never silently slip past this guard (the
    ' Phase-265 Mount omission a hardcoded list allowed).
    Sub Run(h As Harness)
        ' Phase 692 - NodeKind is flat: every kind and the six structural cases
        ' enumerate off the one DU, so the pin is total by reflection alone.
        Dim kindCaseNames = UnionCaseNames(GetType(FsTypes.NodeKind(Of Object)))

        For Each caseName In kindCaseNames
            ' LayoutKind.GridLayout is authored as <Grid> (DataGrid keeps its name).
            Dim expected = If(caseName = "GridLayout", "Grid", caseName)
            h.Check($"coverage: <{expected}> maps kind {caseName}",
                    FuaranXml.IsKnownElement(expected),
                    $"no VB XML element maps kind {caseName}")
        Next
    End Sub

    Private Function UnionCaseNames(duType As Type) As IEnumerable(Of String)
        Return FSharpType.GetUnionCases(duType, Nothing).Select(Function(c) c.Name)
    End Function

End Module

File: src/Fuaran.UI.CSharp.Conformance.Tests/Coverage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.FSharp.Reflection;
// `Fuaran.UI.Types.NodeKind` is a type ABBREVIATION since the 692 swap — erased
// in metadata, so C# reflects the generated declaration directly.
using FsTypes = Fuaran.UI.Generated;

namespace Fuaran.UI.CSharp.Conformance.Tests;

// Phase 305 coverage guard — the §11 forward-coupling reminder for the veneer:
// enumerate every shipped NodeKind/spec case by reflecting the F# DUs and assert a
// public C# factory exists for each. A new F# kind with no C# factory fails here,
// exactly as F#'s FS0025 exhaustiveness would flag a new case.
internal static class Coverage
{
    // The few cases whose C# factory name(s) differ from the F# DU case name. A
    // case maps to the SET of factories that author it, not to one name: `Media`
    // (Phase 1076) is the first kind whose spec carries a VARIANT DU
    // (`MediaKind = Video of VideoDetail | Audio`), and the veneer authors one
    // factory per variant rather than one factory taking a discriminator — video
    // has an autoplay/poster payload audio deliberately has not. Every listed
    // factory is REQUIRED, not any-of: a half-authored variant DU is precisely
    // the gap this pin exists to catch.
    private static readonly IReadOnlyDictionary<string, string[]> Alias = new Dictionary<string, string[]>
    {
        ["GridLayout"] = ["Grid"], // LayoutKind.GridLayout → Fuaran.Grid (DataGrid keeps its name)
        ["Media"] = ["Video", "Audio"], // MediaKind = Video of VideoDetail | Audio
    };

    /// <summary>The C# factory name(s) that author a given wire <c>kind.$type</c>.</summary>
    private static string[] FactoriesFor(string kindType) =>
        Alias.TryGetValue(kindType, out var mapped) ? mapped : [kindType];

    private static readonly HashSet<string> FactoryNames = typeof(Fuaran)
        .GetMethods(BindingFlags.Public | BindingFlags.Static)
        .Where(m => typeof(FuaranNode).IsAssignableFrom(m.ReturnType))
        .Select(m => m.Name)
        .ToHashSet(StringComparer.Ordinal);

    /// <summary>Whether every public C# factory that authors the given wire <c>kind.$type</c> exists.</summary>
    public static bool HasFactoryForKind(string kindType) =>
        FactoriesFor(kindType).All(FactoryNames.Contains);

    public static void Run(Harness h)
    {
        var factoryNames = FactoryNames;

        // Phase 692 — NodeKind is flat: every kind (and the six structural
        // cases) enumerates directly off the one DU, which is exactly what
        // makes this pin total by reflection rather than by four lists.
        var kindCaseNames = UnionCaseNames(typeof(FsTypes.NodeKind<object>)).ToList();

        foreach (var caseName in kindCaseNames)
        {
            var expected = FactoriesFor(caseName);
            var missing = expected.Where(n => !factoryNames.Contains(n)).ToList();
            h.Check(
                $"coverage: factory {string.Join(" + ", expected.Select(n => "Fuaran." + n))} exists for kind {caseName}",
                missing.Count == 0,
                $"no public static {string.Join(" / ", missing.Select(n => $"Fuaran.{n}(…)"))} returning FuaranNode");
        }
    }

    private static IEnumerable<string> UnionCaseNames(Type duType) =>
        FSharpType.GetUnionCases(duType, null).Select(c => c.Name);
}
